### PR TITLE
WIP: dict.jl resize/refactor

### PR DESCRIPTION
**Pull request by chethega**
We can very cheaply clean up after deletions: deleted (`0x02`) slots that are followed by empty (`0x00`) slots can be reset to `0x00`. For speed, this is maintained only approximately (on deletion, we move the head of a `0x02`-chain back by max 4 indices, mostly branch-free).

We can then also maintain a new invariant: `ndel` equals the number of deleted (`0x02`)-slots (if we insert into deleted slot, then decrement `ndel`). 

Using this, we can almost reach homeostasis: Take a dictionary, pop a random item and add a new random item (without changing the size). One rollover is when `count` items have been added and removed. I took some measurements on [https://gist.github.com/chethega/1b11bcb8aa1a6871392cd44e03f4b6e5](https://gist.github.com/chethega/1b11bcb8aa1a6871392cd44e03f4b6e5); upshot: At a fill (`h.count/length(h.slots)`) of 25%, I observed about 18%  deletion (`h.ndel/length(h.slots)`) after 32 rollovers, increasing to 27% after 1024 rollovers. Not a single `rehash` triggered.

Of course, aiming for 1024 rollovers without rehash is overkill. But the current code, rehashing every 3 rollovers (at 0.25 fill) or every 1.5 rollovers (at 0.5 fill) is excessive.

Hence, we should update the entire thresholding logic: The decisions of how much to over-allocate, when to rehash due to deletions, when to shrink (which is perfectly possible).

### CODE REVIEW

**Comment by JeffBezanson** [comment]
Thank you for all the work you're putting in on Dict performance! It's awesome.

File: base/dict.jl [hunk 1]
```diff
@@ -618,6 +618,42 @@ function pop!(h::Dict)
     key => val
 end
 
+@inline function _delete_clean!(h::Dict{K,V}, index) where {K,V}
+    isbitstype(K) || isbitsunion(K) || ccall(:jl_arrayunset, Cvoid, (Any, UInt), h.keys, index-1)
+    isbitstype(V) || isbitsunion(V) || ccall(:jl_arrayunset, Cvoid, (Any, UInt), h.vals, index-1)
+    off = 3
```
Comment (ararslan) [inline base/dict.jl]:
```suggestion
    off = 3
```
should fix CI

**Comment by chethega** [comment]
Ok, sorry for maybe being unclear: the new resize logic does not exist yet (wip, not ready to merge). I am still very unclear about good thresholds (grow at what percentage full? shrink at what percentage full? what is the desired fill rate? when to rehash?). Especially when correctly tracking `ndel` we need to be much more aggressive about rehashing: The old logic guaranteed a rehash every 1.5 rollovers at 75% deleted even if the table is still ok; in the new world, 20% deleted after 30 rollovers is a pretty deterioated dict already, and should be rehashed. 

I think the old resize logic was grown by too many people over too many years and is pretty weird, instead of being consciously designed. But I don't have a good opinion on good resizing logic (needs proper hysteresis!), and lack data on typical use (trade: memory vs time, typical use: insertion/deletion vs lookup/update, how much overallocation at what absolute size and per element size). 

Does anyone here have an opinion on desired levels of overallocation?

**Comment by StefanKarpinski** [comment]
A more coherent, conscious design sounds like a great thing to have—thank your for working towards it. I'm not sure there are right answers for how rehashing thresholds. Some interesting discussion here:

https://stackoverflow.com/questions/2369467/why-are-hash-table-expansions-usually-done-by-doubling-the-size

Doesn't address the fullness thresholds but does address the expansion strategy. My gut says that we should shrink a table if it becomes less than 50% full and grow it if it becomes more than 75% full, but that's based on nothing. I also suspect in my gut that we should maybe grow by a smaller factor than we shrink: e.g. grow by 50% (grow to 3/2x) and shrink by 50% (shrink to 1/2x).

One of the reasons often given for doubling and halving only is that then hash values can be computed with bit masking instead of needing to use the very slow integer. However, we could potentially use a multiply-shift based algorithm and compute the constants involved at hash table resize time. Here's one technique to consider:

https://lemire.me/blog/2019/02/08/faster-remainders-when-the-divisor-is-a-constant-beating-compilers-and-libdivide/

In short, I don't think we necessarily need to stick to power-of-two-sized hash tables.

Another consideration might be adjusing the growth and shrinkage factors based on how much available memory there is. If your hash table is already larger than half of available memory then you definitely don't want to grow it by a factor of two.

**Comment by chethega** [comment]
Currently, we grow a dict at 66% fill, and after growth it has 16% or 33% fill. The general method used (`_tblsz(2*count)` resp `_tblsz(4*count)`)  gives 25%-50% fill, resp 12.5%-25% fill. We currently never shrink. The current `sizehint` gives 50%-100% fill, which predictably leads to a rehash+growth during construction and is therefore pretty useless. 

The threshold for 2x vs 4x growth should take `sizeof(K)+sizeof(V)` into account.
